src/adapt/proto.ts

```diff
@@ -0,0 +1,499 @@
+  let fdpName;
+  if (fds.file.length > 0) {
+    // search root descriptor
+    const fdp = fds.file[0];
+    fdpName = fdp.name;
+    if (fdp.messageType && fdp.messageType.length > 0) {
+      dp = new DescriptorProto(fdp.messageType[0]);
+    }
+  }
+  if (!dp) {
+    throw Error('root descriptor not found');
+  }
+  for (const fdp of fds.file) {
+    if (fdp.name === fdpName) {
+      continue;
+    }
+    if (!dp.nestedType) {
+      dp.nestedType = [];
+    }
+    if (!fdp.messageType) {
+      continue;
+    }
+    for (const nestedDP of fdp.messageType) {
+      dp.nestedType.push(
+        normalizeDescriptorProto(new DescriptorProto(nestedDP))
+      );
+    }
+  }
+  return normalizeDescriptorProto(dp);
+}
+
+function normalizeDescriptorProto(dp: DescriptorProto): DescriptorProto {
+  dp.name = normalizeName(dp.name);
+  for (const f of dp.field) {
+    if (f.proto3Optional) {
+      f.proto3Optional = null;
+    }
+    if (f.oneofIndex) {
+      f.oneofIndex = null;
+    }
+  }
+  return dp;
+}
+
+function normalizeName(name: string): string {
+  return name.replace(/\./, '_');
+}
+
+/** Builds a namespace descriptor in JSON format from a FileDescriptorSet.
+ * This can be used to on protobufjs.Root.fromJSON to load protobuf structs.
+ * @param fds - a FileDescriptorSet
+ * @returns Namespace
+ */
+export function fileDescriptorSetToNamespace(
+  fds: FileDescriptorSet
+): Namespace {
+  const ns: Namespace = {
+    nested: {},
+  };
+  for (const fdp of fds.file || []) {
+    for (const dp of fdp.messageType || []) {
+      if (!ns.nested) {
+        ns.nested = {};
+      }
+      const data = protoDescriptorToNamespace(new DescriptorProto(dp));
+      for (const key of Object.keys(data.nested)) {
+        ns.nested[`${key}`] = data.nested[key];
+      }
+    }
+  }
+  return ns;
+}
+
+/** Builds a namespace descriptor in JSON format from a DescriptorProto.
+ * This can be used to on protobufjs.Root.fromJSON to load protobuf structs.
+ * @param dp - a DescriptorProto
+ * @returns Namespace
+ */
+export function protoDescriptorToNamespace(dp: DescriptorProto): Namespace {
+  const fields: {[k: string]: protobuf.IField} = {};
+  for (const f of dp.field || []) {
+    fields[`${f.name}`] = descriptorProtoFieldToField(f);
+  }
+  const result = {
+    nested: {
+      [`${dp.name}`]: {
+        fields,
+      },
+    },
+  };
+  for (const nested of dp.nestedType || []) {
+    const nestedData = protoDescriptorToNamespace(new DescriptorProto(nested));
+    for (const key of Object.keys(nestedData.nested)) {
+      const data = nestedData.nested[key];
+      if (data) {
+        result.nested[key] = data;
+      }
+    }
+  }
+  return result;
+}
+
+function descriptorProtoFieldToField(
+  field: FieldDescriptorProto
+): protobuf.IField {
+  let type = '';
+  if (field.type) {
+    if (field.typeName) {
+      type = field.typeName;
+    } else {
+      type = fieldTypeLabelMap[field.type].replace('TYPE_', '').toLowerCase();
```

Comment:
I was reminded the other day, in case it matters, that replace(string, string) will only replace the first occurrence. You'd need to use `/TYPE_/g` to get all of them. (not sure if that applies here)